### PR TITLE
Plots

### PR DESCRIPTION
**Pull request by dmpetrov**
Plot metrics.

Related to https://github.com/iterative/dvc/issues/3409

### CODE REVIEW

**Comment by dmpetrov** [comment]
@jorgeorpinel and @shcheklein for some reason I don't see `dvc plot` & `dvc plot show\diff` in the menu. It looks like I missed something.
Could you please help me?

**Comment by shcheklein** [comment]
@dmpetrov yep, update sidebar.json - that should do the trick.

**Comment by dmpetrov** [comment]
> sidebar

Thank you @shcheklein ! Forgot about this :)

File: content/docs/command-reference/plot/index.md [hunk 1]
```diff
@@ -0,0 +1,188 @@
+# plot
+
+Contains commands to visualize users data stored in structured files like JSON,
+CSV, TSV: [show](/doc/command-reference/plot/show).
```
Comment (jorgeorpinel) [inline content/docs/command-reference/plot/index.md:4]:
```suggestion
Contains commands to visualize
[continuous metrics](/doc/command-reference/plot#continuous-metrics) in
structured files like JSON, CSV, TSV:
[show](/doc/command-reference/plot/show),
```

File: content/docs/command-reference/plot/index.md [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+[diff](/doc/command-reference/plot/diff).
+
+## Synopsis
+
+```usage
+usage: dvc plot [-h] [-q | -v] {show,diff} ...
+positional arguments:
+  {show,diff}    Use `dvc plot CMD --help` to display command-specific help.
+    show         Plot data from a file
```
Comment (jorgeorpinel) [inline content/docs/command-reference/plot/index.md:13]:
```suggestion
usage: dvc plot [-h] [-q | -v] {show,diff} ...

positional arguments:
  COMMAND
    show         Plot data from a file
```

File: content/docs/command-reference/plot/index.md [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+                 or between the last commit and the workspace.
+```
+
+## Description
+
+DVC provides a set of commands to visualize
+[continuous metrics](/doc/command-reference/plot#continous-metrics) of machine
+learning experiments in <abbr>DVC projects</abbr>. These metrics are usually
+plotted as AUC curves, loss functions, confusion matrixes and other types of
+plots.
```
Comment (jorgeorpinel) [inline content/docs/command-reference/plot/index.md:24]:
```suggestion
DVC provides a set of commands to visualize continuous metrics of machine
learning experiments in. Usual examples of plots are AUC curves, loss
functions, and confusion matrices.
```

File: content/docs/command-reference/plot/index.md [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+The continuous metrics should be saved in files which are usually created by
+users or generated by user's modeling or data processing code. The plot commands
+can work with these files commited to a repository history, data files
+controlled by DVC or files from workspace.
+
+### Continous metrics
```
Comment (jorgeorpinel) [inline content/docs/command-reference/plot/index.md:31]:
```suggestion
### Continuous metrics
```

File: content/docs/command-reference/plot/index.md [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+DVC has two concepts for metrics for representing result of machine learning
+training or data processing:
+
+1. `dvc metrics` to represent scalar numbers such as AUC, true positive rate and
+   others.
+2. `dvc plot` to visualize continuous metrics such as AUC curve, loss function,
+   confusion matrixes and others.
```
Comment (jorgeorpinel) [inline content/docs/command-reference/plot/index.md:39]: Question: maybe we should use "scalar metrics" everywhere i.e. in the dvc metrics commands and other places where they're currently mentioned? Always try to use either scalar or continuous except in placs where metrics are mentioned in an abstract sense.

Comment (dmpetrov) [reply content/docs/command-reference/plot/index.md:39]: > maybe we should use "scalar metrics" everywhere

Yes. It should be a very clear separation.

**Comment by dmpetrov** [comment]
> Please let us know when this is ready for full review.

@jorgeorpinel  thank you for the small changes.

I'm waiting for a bugfix while I can add one more part about the plot templates. It would be great to make review now. I'll incorporate the template later right here or as a separate PR.

File: content/docs/command-reference/plot/diff.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+- `-q`, `--quiet` - do not write anything to standard output. Exit with 0 if no
+  problems arise, otherwise 1.
+
+- `-v`, `--verbose` - displays detailed tracing information.
+
+## Examples
+
+The difference between a not commited version of the file and the last commited
+one:
```
[the remark this answers is not in the document]
Comment (dmpetrov) [reply content/docs/command-reference/plot/diff.md:73]: provided a additional sentence.

File: content/docs/command-reference/plot/show.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
+# plot show
+
+Generate a [plot](/doc/command-reference/plot) from a file with
+[continuous metrics](/doc/command-reference/plot#continous-metrics).
+
+## Synopsis
+
+```usage
+usage: dvc plot show [-h] [-q | -v] [-t [TEMPLATE]] [-r RESULT] [--show-json]
```
[the remark this answers is not in the document]
Comment (dmpetrov) [reply content/docs/command-reference/plot/show.md:9]: right

**Comment by dmpetrov** [comment]
@jorgeorpinel and @pared pushed all the changes. Please take a look.

File: content/docs/command-reference/plot/index.md (hunk 5 above)
Comment (jorgeorpinel) [reply content/docs/command-reference/plot/index.md:39]: OK added a checkbox to #1175

File: content/docs/command-reference/plot/index.md [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+positional arguments:
+  COMMAND
+    show         Plot data from a file
+    diff         Plot changes between commits in the DVC repository,
+                 or between the last commit and the workspace.
+```
+
+## Description
+
+DVC provides a set of commands to visualize continuous metrics of machine
```
Comment (jorgeorpinel) [inline content/docs/command-reference/plot/index.md:22]:
```suggestion
DVC provides a set of commands to visualize _continuous metrics_ of machine
```

File: content/docs/command-reference/plot/index.md [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+34, 0.91935, 0.0317345
+35, 0.91913, 0.0317829
+36, 0.92256, 0.0304632
+37, 0.92302, 0.0299015
+```
+
+In hierarchical file formats such as JSON an array of JSON-objects is expected.
+Plot command can generate visuals for a specified field name or a set of fields
+from the array's object. Like `val_loss` field in the `train` array in this
+example:
```
Comment (jorgeorpinel) [inline content/docs/command-reference/plot/index.md:84]:
```suggestion
In JSON files (hierarchical format) a named array of matching JSON objects is
expected. `dvc plot show` can generate visuals for a one, several, or all field
names from the array's objects. For example `val_loss` in the `train` array below:
```

File: content/docs/command-reference/plot/show.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
+# plot show
+
+Generate a [plot](/doc/command-reference/plot) from a file with
+[continuous metrics](/doc/command-reference/plot#continous-metrics).
```
Comment (jorgeorpinel) [inline content/docs/command-reference/plot/show.md:4]:
```suggestion
Generate a plot image file from a
[continuous metrics](/doc/command-reference/plot#continous-metrics) file.
```

File: content/docs/command-reference/plot/show.md [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+                     [datafile]
+
+positional arguments:
+  datafile              Data to be visualized.
+```
+
+## Description
+
+This command provides a quick way to visualize countinuous metrics such as loss
+functions, AUC curves, confusion matrixes etc. Please read `dvc params` for more
```
Comment (jorgeorpinel) [inline content/docs/command-reference/plot/show.md:20]:
```suggestion
functions, AUC curves, confusion matrixes etc. Please read `dvc plot` for more
```

File: content/docs/command-reference/plot/diff.md [hunk 12]
```diff
@@ -0,0 +1,115 @@
+# plot diff
+
+Show difference in
+[continuous metrics](/doc/command-reference/plot#continous-metrics) by plotting
+on a single [plot](/doc/command-reference/plot) different versions of metrics
+from the <abbr>DVC repository</abbr> or workspace.
```
Comment (jorgeorpinel) [inline content/docs/command-reference/plot/diff.md:6]: I would simplify like this:
```suggestion
different versions in a single image.
```
and leave the other details for the description.

File: content/docs/command-reference/plot/show.md [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+## Synopsis
+
+```usage
+usage: dvc plot show [-h] [-q | -v] [-t [TEMPLATE]] [-r RESULT] [--show-json]
+                     [-f FIELDS]
+                     [datafile]
+
+positional arguments:
+  datafile              Data to be visualized.
```
Comment (jorgeorpinel) [inline content/docs/command-reference/plot/show.md:14]:
```suggestion
  datafile              Continuous metrics file to visualize
```
Cc @pared for the core repo PR

p.s. should this term use singular or plural "metric(s)" ?

File: content/docs/command-reference/plot/index.md [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+## Synopsis
+
+```usage
+usage: dvc plot [-h] [-q | -v] {show,diff} ...
+
+positional arguments:
+  COMMAND
+    show         Plot data from a file
+    diff         Plot changes between commits in the DVC repository,
+                 or between the last commit and the workspace.
```
Comment (jorgeorpinel) [inline content/docs/command-reference/plot/index.md:17]:
```suggestion
    show         Generate a plot image file from a continuous metrics file
    diff         Plot continuous metrics differences between commits in the DVC repository,
                 or between the last commit and the workspace.
```
> May need rewrapping
Cc @pared also for the core repo PR

File: content/docs/command-reference/plot/show.md (hunk 13 above)
Comment (jorgeorpinel) [reply content/docs/command-reference/plot/show.md:14]: Moved to iterative/dvc/pull/3577

File: content/docs/command-reference/plot/index.md (hunk 14 above)
Comment (jorgeorpinel) [reply content/docs/command-reference/plot/index.md:17]: Moved to iterative/dvc/pull/3577

File: content/docs/command-reference/plot/index.md [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+```
+
+## Description
+
+DVC provides a set of commands to visualize continuous metrics of machine
+learning experiments in. Usual examples of plots are AUC curves, loss functions,
+and confusion matrices.
```
[the remark this answers is not in the document]
Comment (dmpetrov) [reply content/docs/command-reference/plot/index.md:24]: All points are great! I rearranged this part. Please take a look.

It is still important to keep the scalars vs continuous difference somewhere. I keep it as a next section but rearranged from the plot point of view.

File: content/docs/command-reference/plot/show.md (hunk 13 above)
Comment (dmpetrov) [reply content/docs/command-reference/plot/show.md:14]: metrics - plural in many docs. let's use plural.

**Comment by dmpetrov** [comment]
@jorgeorpinel thank you for the review. Please take a look at the changes.

File: content/docs/command-reference/plot/index.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply content/docs/command-reference/plot/index.md]: Thanks! I think it's easier to read now 🙂 

> image

I agree! Unfortunately I pasted the wrong text in my previous message and so I didn't express myself correctly haha. This is what I meant:

image

So to summarize, I think the scalar metrics should be explained in the description of `dvc metrics` cmd ref index, and just mention the word here with a link to there, without explaining. But OK I guess that can be a separate issue (to match dvc metrics with this ref)

[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply content/docs/command-reference/plot/index.md]: Oops I did it again (wrong paste). Please read the comment above in Github ☝️ because I edited it...

[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply content/docs/command-reference/plot/index.md]: p.s. unresolving for myself to remember about "that can be a separate issue (to match dvc metrics with this ref)"

File: content/docs/command-reference/plot/show.md (hunk 13 above)
Comment (jorgeorpinel) [reply content/docs/command-reference/plot/show.md:14]: This has been applied in the core repo, so we can do it here too 🙂

File: content/docs/command-reference/plot/index.md (hunk 14 above)
Comment (jorgeorpinel) [reply content/docs/command-reference/plot/index.md]: The definition for show is still pending in the core repo (https://github.com/iterative/dvc/pull/3577#discussion_r416844217)
But the one for diff has been applied ✅

Comment (jorgeorpinel) [reply content/docs/command-reference/plot/index.md]: Oh I see this one has been applied. Thanks, resolving.

File: content/docs/command-reference/plot/index.md [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+    show         Generate a plot image file from a continuous metrics file.
+    diff         Plot continuous metrics differences between commits in the DVC
+                 repository, or between the last commit and the workspace.
+```
+
+## Description
+
+DVC provides a set of commands to visualize _continuous metrics_ of machine
+learning experiments. Usual examples of plots are AUC curves, loss functions,
+and confusion matrices.
```
Comment (jorgeorpinel) [inline content/docs/command-reference/plot/index.md:24]:
```suggestion
Continuous metrics represent plots, and should be stored as data series in one
of the supported [file formats](#file-formats). These files are usually created
by users or generated by user modeling or data processing code.
```

File: content/docs/command-reference/plot/index.md [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+
+Continous metrics represents a plot and should be stored as data series in one
+of the supported [file formats](#file-formats). These files are usually created
+by users or generated by user's modeling or data processing code.
```
Comment (jorgeorpinel) [inline content/docs/command-reference/plot/index.md:28]:
```suggestion
Continuous metrics represent plots, and should be stored as data series in one
of the supported [file formats](#file-formats). These files are usually created
by users or generated by user modeling or data processing code.
```

Comment (jorgeorpinel) [reply content/docs/command-reference/plot/index.md:28]: But how are plots defined? Is there a new option in `dvc run` for these, like -m for metrics? I don't see info about this in any plot ref yet, nor an update on run itself. Does that mean these metrrics are not tracked by DVC like scalar metrics? Probably worth mentioning either way.

**Comment by jorgeorpinel** [comment]
> thank you for the review. Please take a look at the changes

No problem. I resolved all the parts of the review that are now addressed or no longer relevant, and added one more comment. Still haven't reviewed everything in detail (options, examples) is it a good time to do so? And again, I can take this over at some point if needed, lmk guys cc @ivan

**Comment by shcheklein** [comment]
@jorgeorpinel wrong `@ ivan` I think :)

what is the status of this?

File: content/docs/command-reference/plot/index.md (hunk 17 above)
Comment (dmpetrov) [reply content/docs/command-reference/plot/index.md:28]: @jorgeorpinel there is no plot metrics in `dvc run` yet. we should introduce one. `-p` is taken by params. What do you think about `-l`\`-L` (pLot)?

Comment (jorgeorpinel) [reply content/docs/command-reference/plot/index.md:28]: I like `-l` (`--plots`) since it comes after `-m` (`--metrics`)

Will there also be a upper case `-L` (`--plot-no-cache`)?

All of this should probably be its own issue in the core repo 😬 I can create it if you'd like, please lmk.

**Comment by jorgeorpinel** [comment]
> wrong @ ivan

I keep doing that...

> what is the status of this?

Some review items are still pending. Are we taking it over?

**Comment by shcheklein** [comment]
@jorgeorpinel I trust you on making this call :) don't have enough context on this.

**Comment by dmpetrov** [comment]
@shcheklein and @jorgeorpinel I just updated it:
1. made all the asked changes
2. added new plot options
3. small API fixes (output file name)
4. added custom plots section

Please take a look,

File: content/docs/command-reference/plot/diff.md [hunk 18]
```diff
@@ -0,0 +1,130 @@
+# plot diff
+
+Show multiple versions of
+[continuous metrics](/doc/command-reference/plot#continous-metrics) by plotting
```
Comment (pared) [inline content/docs/command-reference/plot/diff.md:4]: I wonder whether we want to name them `continuous`. This word applies to functions. What about, for example, confusion matrix? Data for that type of plot is not continuous.

File: content/docs/command-reference/plot/index.md [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+the metrics:
+
+```dvc
+$ dvc plot diff -d logs.csv
+file:///Users/dmitry/src/plot/logs.html
+```
+
+![](/img/plot_auc.svg)
+
+### Difference between continuous and scolar metrics
```
Comment (pared) [inline content/docs/command-reference/plot/index.md:42]: scalar

File: content/docs/command-reference/plot/show.md (hunk 13 above)
Comment (pared) [reply content/docs/command-reference/plot/show.md:14]: I don't like `continuous`, my doubts are summarized in my first comment.

File: content/docs/command-reference/plot/diff.md (hunk 18 above)
Comment (shcheklein) [reply content/docs/command-reference/plot/diff.md:4]: Agreed with @pared . Probably even plain explicit "non-scalar metric" would be better?

Comment (dmpetrov) [reply content/docs/command-reference/plot/diff.md:4]: yeah. it will be changed. all the terminology around continuous will be removed in the next iteration.

File: content/docs/command-reference/plot/show.md (hunk 13 above)
Comment (jorgeorpinel) [reply content/docs/command-reference/plot/show.md:14]: Agree, but that seems to be the paragraph in the core repo right now so I'm going to commit this for now. We can update it again when needed.

Comment (jorgeorpinel) [reply content/docs/command-reference/plot/show.md:14]: Ah never mind I see its `Metrics file to visualize.` now. Will update 🙂

Comment (jorgeorpinel) [reply content/docs/command-reference/plot/show.md:14]: Btw extra period 😅

Comment (jorgeorpinel) [reply content/docs/command-reference/plot/show.md]: p.s. please see iterative/dvc/pull/3759 @pared

File: content/docs/command-reference/plot/diff.md (hunk 18 above)
Comment (jorgeorpinel) [reply content/docs/command-reference/plot/diff.md:4]: - [ ] So let's leave "continuous" in this PR so we have a plot cmd ref for now, and update it in a following PR.

File: content/docs/command-reference/plot/index.md (hunk 19 above)
Comment (jorgeorpinel) [reply content/docs/command-reference/plot/index.md:42]: Adding suggestion for quick commit from GH:
```suggestion
### Difference between continuous and scalar metrics
```
